### PR TITLE
OP-34 Add Testing Capability

### PR DESCRIPTION
**Pull request by jassol**
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

[//]: # "Describe what's new in this PR"

Added node modules to enable testing with Mocha/Chai/Sinon in the dev environment
### How to Review
Main changes are adding a folder for tests, and the first test file for billGeneration.

[//]: # 'The order in which to review files and what to expect when testing locally'

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc"

### Next steps

[//]: # "What doesn't work yet, what's NOT in this PR/has to be done "

### Tests Performed, Edge Cases

[//]: # 'If you made changes/additions to `airtable.js` make sure you update/add'
[//]: # '[at least 1] corresponding unit tests in `airtable.spec.js`'

### Screenshots

[//]: # "Add screenshots!!! If you'd like"

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -7,9 +7,9 @@ import {
   approveSubscriberBill
 } from './utils/billgeneration';
 import { sendInviteEmail } from './utils/pledgeInvite';
-import generatePdfForSubscriber from './utils/pdfgeneration';
+import {generatePdfForSubscriber} from './utils/pdfgeneration';
```
Comment (jassol) [inline index.js:10]: Destructured import/export to enable Sinon stubbing

File: index.js [hunk 2]
```diff
@@ -7,9 +7,9 @@ import {
...
 import EmailGenerators from './utils/emailCopy';
-import sendEmail from './utils/email';
+import { sendEmail } from './utils/email';
```
Comment (jassol) [inline index.js:12]: Destructured import/export to enable Sinon stubbing

File: package.json [hunk 3]
```diff
@@ -26,9 +26,15 @@
   },
   "scripts": {
     "generate-schema": "generate-airtable-schema && pretty-quick",
-    "start": "babel-node index.js"
+    "start": "babel-node index.js",
+    "test": "mocha --require @babel/register --require babel-polyfill --recursive --file test/billGenerationTests.test.js"
```
Comment (jassol) [inline package.json:30]: Added test script. Babel/register and babel/polyfill are necessary for compiling ES6. The recursive tag will be needed when we ad another test file. For now, I am only running the one test file.

File: utils/billgeneration.js [hunk 4]
```diff
@@ -11,10 +11,10 @@ import {
 } from '../airtable/request';
 import { getEnphaseDataForSubscriber } from './enphase';
 
-import sendEmail from './email';
+import { sendEmail } from './email';
```
Comment (jassol) [inline utils/billgeneration.js:14]: Destructured import/export to enable Sinon stubbing

File: utils/billgeneration.js [hunk 5]
```diff
@@ -11,10 +11,10 @@ import {
...
 import EmailGenerators from './emailCopy';
-import getLatestPGEBill from './utilityApi';
-import generatePdfForSubscriber from './pdfgeneration';
+import { getLatestPGEBill } from './utilityApi';
+import { generatePdfForSubscriber } from './pdfgeneration';
```
Comment (jassol) [inline utils/billgeneration.js:17]: Destructured import/export to enable Sinon stubbing

File: utils/email.js [hunk 6]
```diff
@@ -26,4 +26,4 @@ const sendEmail = async email => {
   return sentEmailRecipient;
 };
 
-export default sendEmail;
+export { sendEmail };
```
Comment (jassol) [inline utils/email.js:29]: Destructured import/export to enable Sinon stubbing

File: utils/pdfgeneration.js [hunk 7]
```diff
@@ -160,4 +160,4 @@ const generatePdfForSubscriber = async (
   }, Constants.PDF_DELETE_DELAY * 1000);
 };
 
-export default generatePdfForSubscriber;
+export {generatePdfForSubscriber};
```
Comment (jassol) [inline utils/pdfgeneration.js:163]: Destructured import/export to enable Sinon stubbing

File: utils/pledgeInvite.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { getPledgeInviteById } from '../airtable/request';
 import EmailGenerators from './emailCopy';
 import Constants from '../Constants';
-import sendEmail from './email';
+import { sendEmail } from './email';
```
Comment (jassol) [inline utils/pledgeInvite.js:4]: Destructured import/export to enable Sinon stubbing

File: utils/utilityApi.js [hunk 9]
```diff
@@ -133,4 +133,4 @@ const getLatestPGEBill = async meterId => {
   };
 };
 
-export default getLatestPGEBill;
+export {getLatestPGEBill};
```
Comment (jassol) [inline utils/utilityApi.js:136]: Destructured import/export to enable Sinon stubbing

File: index.js (hunk 2 above)
Comment (jassol) [inline index.js:12]: Destructured import/export to enable Sinon stubbing

File: index.js (hunk 1 above)
Comment (grflood) [inline index.js:10]: add spaces before/after brackets?

File: test/billGenerationTests.test.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  approveSubscriberBill,
+  generateBillsForSolarProject
+} from '../utils/billgeneration';
+
+describe('billGeneration', () => {
+  const sandbox = sinon.createSandbox();
+
+  describe('#generateBillsForSolarProject', async () => {
+
+    const subscriberThree = {id: 3, meterId: 2, name: 'Lo', subscriberBillIds: [5, 2], rateScheduleId: 2};
```
Comment (grflood) [inline test/billGenerationTests.test.js:21]: maybe add a `TODO OP-24: Make stock test objects for tests` ahead of all this ?

File: test/billGenerationTests.test.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    const sixtyDaysAgo = moment().subtract(60, 'days');
+    const solarProjectId = 7;
+    const solarProject = {
+      id: 7,
+      subscriberIds: [3,5,9],
+      name: "Magical Solar Extravaganza",
+      enphaseUserId: 17,
+      enphaseSystemId: 21
+    };
+    // TODO for all PGE mock bills: what units should these be? use decimals?
```
Comment (grflood) [inline test/billGenerationTests.test.js:35]: we should mirror whatever data type is actually sent to this code from Utility API and our code

File: test/billGenerationTests.test.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    const rateScheduleForSubNine = {id: 4, rebateRate: 1.002, rate: 0.06};
+
+
+    before(async () => {
+      sandbox.stub(request, 'getSolarProjectById').returns(solarProject);
+      const stubbedGetOwnerById = sandbox.stub(request, 'getOwnerById');
+      stubbedGetOwnerById.withArgs(3).returns(subscriberThree);
+      stubbedGetOwnerById.withArgs(5).returns(subscriberFive);
+      stubbedGetOwnerById.withArgs(9).returns(subscriberNine);
+      const stubbedGetPGEBill = sandbox.stub(utilityAPI, 'getLatestPGEBill');
```
Comment (grflood) [inline test/billGenerationTests.test.js:86]: newlines to break up the different stubs would be great, one before each new `const` is declared

File: test/billGenerationTests.test.js [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+      stubbedGetPGEBill.withArgs(2).returns(latestPGEBillMeterTwo);
+      stubbedGetPGEBill.withArgs(8).returns(latestPGEBillMeterEight);
+      stubbedGetPGEBill.withArgs(7).returns(latestPGEBillMeterSeven);
+      const stubbedGetSubBullsById = sandbox.stub(request, 'getSubscriberBillsByIds')
```
Comment (grflood) [inline test/billGenerationTests.test.js:90]: Bulls -> Bills =P

File: test/billGenerationTests.test.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+      stubbedGetEnphaseData.onCall(1).returns(generationDataForSubFive);
+      stubbedGetEnphaseData.onCall(2).returns(generationDataForSubNine);
+      const stubbedGetRateSchedule = sandbox.stub(request, 'getRateScheduleById');
+      stubbedGetRateSchedule.withArgs(2).returns(rateScheduleForSubThree);
+      stubbedGetRateSchedule.withArgs(3).returns(rateScheduleForSubFive);
+      stubbedGetRateSchedule.withArgs(4).returns(rateScheduleForSubNine);
+      sandbox.stub(request, 'createSubscriberBill');
+      sandbox.stub(pdfgeneration, 'generatePdfForSubscriber');
+      sandbox.stub(email, 'sendEmail'); // TODO maybe should throw an error for test
+      await generateBillsForSolarProject(solarProjectId);
```
Comment (grflood) [inline test/billGenerationTests.test.js:107]: imo it's not a good idea to put the function call in the sandbox set up code. In the future when you test different cases and scenarios (for example a case where the function returns an error) you'll want to modify this default sandbox set up before calling `generateBillsForSolarProject`

File: test/billGenerationTests.test.js [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+      sandbox.restore();
+    });
+
+    it("should get a solar project by its id", async () => {
+      expect(request.getSolarProjectById.called).to.equal(true);
+    });
+
+    it("should get each subscriber associated with the solar project", async () => {
+      expect(request.getOwnerById.called).to.equal(true);
+      expect(request.getOwnerById.calledWith(3)).to.equal(true);
```
Comment (grflood) [inline test/billGenerationTests.test.js:120]: small detail, but changing this and future statements to say something like:

```      expect(request.getOwnerById.calledWith(subscriberThree.id)).to.equal(true);```

where you use the name of the variable instead of the integer helps with readability, it's not immediately apparent what the integers are meant to refer to

File: test/billGenerationTests.test.js [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+      expect(enphase.getEnphaseDataForSubscriber.getCall(0).args[0]).to.equal(17);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(0).args[1]).to.equal(21);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(0).args[4]).to.equal(3);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(1).args[0]).to.equal(17);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(1).args[4]).to.equal(5);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(2).args[0]).to.equal(17);
+      expect(enphase.getEnphaseDataForSubscriber.getCall(2).args[4]).to.equal(9);
+    });
+
+    // tests to add:
```
Comment (grflood) [inline test/billGenerationTests.test.js:156]: if you could leave a TODO with a ticket # so we know it's tracked

File: test/billGenerationTests.test.js [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+    ];
+    const generationDataForSubThree = [5263, 4112, 6023, 4801];
+    const generationDataForSubFive = [9873, 9827, 3648, 8642];
+    const generationDataForSubNine = [1253, 1911, 2144, 1306];
+    const rateScheduleForSubThree = {id: 2, rebateRate: 0.093, rate: 0.18};
+    const rateScheduleForSubFive = {id: 3, rebateRate: 0.048, rate: 0.21};
+    const rateScheduleForSubNine = {id: 4, rebateRate: 1.002, rate: 0.06};
+
+
+    before(async () => {
```
Comment (grflood) [inline test/billGenerationTests.test.js:80]: so many stubs but well done !!

Comment (grflood) [reply test/billGenerationTests.test.js:80]: just a monster function =/ sry this was the first one u chose lol

File: utils/pdfgeneration.js (hunk 7 above)
Comment (grflood) [reply utils/pdfgeneration.js:163]: spaces before/after brackets =P

File: utils/utilityApi.js (hunk 9 above)
Comment (grflood) [reply utils/utilityApi.js:136]: spaces again
